### PR TITLE
Adding SSL support to db_redis and ndb_redis

### PR DESCRIPTION
**Pull request by joelbax**
#### Pre-Submission Checklist

- [x] Commit message has the format required by CONTRIBUTING guide
- [x] Commits are split per component (core, individual modules, libs, utils, ...)
- [x] Each component has a single commit (if not, squash them into one commit)
- [x] No commits to README files for modules (changes must be done to docbook files
in `doc/` subfolder, the README file is autogenerated)

#### Type Of Change
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)

#### Checklist:

- [ ] PR should be backported to stable branches
- [x] Tested changes locally
- [ ] Related to issue #XXXX (replace XXXX with an open issue number)

#### Description
A couple of commit (1 per each module) adding SSL support to db_redis and ndb_redis. 

This mainly includes checking if proper parameter is provided (for ndb_redis it is `ssl` option in the DB URL and, for db_redis, a new `opt_ssl` parameter) and create a temporary SSL context that is used to initialise the redis context.

db_redis is also updated with another parameter to provide a DB access password.

### CODE REVIEW

File: src/modules/db_redis/redis_connection.c [hunk 1]
```diff
@@ -137,14 +140,23 @@ int db_redis_connect(km_redis_con_t *con) {
...
-    } else {
-        LM_ERR("invalid url scheme\n");
+    } 
+    
+    if (db_redis_opt_ssl != 0) {
+        /* Create SSL context*/
+	redisInitOpenSSL();
+	ssl = redisCreateSSLContext(NULL, NULL, NULL, NULL, NULL, NULL);
+	if (ssl == NULL) {
+	    LM_ERR("Unable to create Redis SSL Context.\n");
```
Comment (henningw) [inline src/modules/db_redis/redis_connection.c:155]: It seems indention here is wrong, please adapt it for the goto and the one bracket.

File: src/modules/db_redis/redis_connection.c [hunk 2]
```diff
@@ -137,14 +140,23 @@ int db_redis_connect(km_redis_con_t *con) {
     char hosts[MAX_URL_LENGTH];
     char* host_begin;
     char* host_end;
+    redisSSLContext *ssl = NULL;
```
Comment (henningw) [inline src/modules/db_redis/redis_connection.c:143]: consider moving this parameter definition to the beginning of the function, as its needed in #if and #else

File: src/modules/db_redis/redis_connection.c [hunk 3]
```diff
@@ -163,13 +175,28 @@ int db_redis_connect(km_redis_con_t *con) {
...
+    if (ssl) {
+        redisClusterSetOptionEnableSSL(con->con, ssl);
+    }
     status = redisClusterConnect2(con->con);
     if (status != REDIS_OK) {
         LM_ERR("cannot open connection to cluster with hosts: %s, error: %s\n", hosts, con->con->errstr);
         goto err;
     }
 #else
+    redisSSLContext *ssl = NULL;
```
Comment (henningw) [inline src/modules/db_redis/redis_connection.c:187]: move to beginning, as before

File: src/modules/db_redis/redis_connection.c [hunk 4]
```diff
@@ -180,10 +207,18 @@ int db_redis_connect(km_redis_con_t *con) {
...
 #endif
 
-    if (con->id->password) {
-        reply = redisCommand(con->con, "AUTH %s", con->id->password);
+    password = con->id->password;
+    if (!password) {
+        password = db_pass;
+    }
+    if (password) {
+        LM_DBG("Using password %s\n", password);
```
Comment (henningw) [inline src/modules/db_redis/redis_connection.c:220]: Please not log the password to the system log, this should be removed.

File: src/modules/ndb_redis/redis_client.h [hunk 5]
```diff
@@ -72,6 +74,7 @@ typedef struct redisc_server {
 	param_t *attrs;
 	char *spec;
 	redisContext *ctxRedis;
+    redisSSLContext *sslCtxRedis;
```
Comment (henningw) [inline src/modules/ndb_redis/redis_client.h:77]: Fix the indention

**Review by henningw — COMMENTED** [review]
Thanks for the PR. I have added a few comments related to white space changes and variable declaration. Please remove the debugging comment regarding the password, its not a good security practice to log it.
For merging this to our code base, you also need to add documentation to the modules. This is done by extending the XML file in the doc directory. They are are used to create the README and web site documentation.

**Comment by joelbax** [comment]
@henningw  Thank you very much for the feedback. 

I've added documentation information into db_redis_admin.xml and ndb_redis_admin.xml files but I was wondering if I needed to add any reference in the authors/editors section in db_redis.xml and ndb_redis.xml. Please, let me know if that's needed.

**Review by henningw — COMMENTED** [review]
Thank you, looks good to me now. There seems to be a build error for gcc and clang related to the hiredis-ssl header in the automatic build tests. Could you have a quick look to that? You can find the build logs by clicking to the failed two jobs, thanks. Maybe we just need to add some packaging rules or similar.

File: src/modules/db_redis/doc/db_redis_admin.xml [hunk 6]
```diff
@@ -213,6 +213,43 @@ modparam("db_redis", "keys", "version=entry:table_name;location=entry:ruid&amp;u
...
 modparam("db_redis", "verbosity", 0)
+...
+				</programlisting>
+			</example>
+		</section>
+
+		<section id="db_redis.p.opt_ssl">
+			<title><varname>opt_ssl</varname> (int)</title>
+			<para>
+				Controls SSL usage while connecting to a remote DB.
```
Comment (oej) [inline src/modules/db_redis/doc/db_redis_admin.xml:224]: Please do not use "SSL" in any documentation as that protocol is old and deemed insecure. Replace with "TLS".

Setting names in new code should reflect this, regardless of the names used by old libraries.

**Comment by joelbax** [comment]
> Thank you, looks good to me now. There seems to be a build error for gcc and clang related to the hiredis-ssl header in the automatic build tests. Could you have a quick look to that? You can find the build logs by clicking to the failed two jobs, thanks. Maybe we just need to add some packaging rules or similar.

I'm really sorry @henningw I totally missed that. I had to build hiredis with SSL support because default Debian packages didn't provide it. SSL is natively supported by hiredis, it was just a matter of getting the source package and building again with USE_SSL=1, I don't know why that's not provided by libhiredis-dev package by default.

Should I reject this PR until libhiredis-dev provides such support?

**Comment by oej** [comment]
Thank you for removing "SSL" !

**Comment by henningw** [comment]
@joelbax Thanks for the update. No worries, for that we are having the automatic build tests. Regarding merging this PR, could you please check if there is maybe a C define that you could check if the SSL is included in the library? Then we could merge it and it should be available as soon its in debian stable etc..

**Comment by henningw** [comment]
@joelbax Any update regarding the last discussed topic about the define?

**Comment by joelbax** [comment]
@henningw  I'm really sorry for the long long delay but I've been constantly side-pulled in work and I couldn't find a define I could use, so, I'm thinking about a different approach checking if the SSL lib exists and do that define in the Makefile.

**Comment by henningw** [comment]
@joelbax Thanks for the reply. Do you have access for a locally compiled and installed libhiredis? Could you maybe execute this command: `$ pkg-config hiredis --cflags`

I am interested if maybe the pkg-config is usable here in deciding if its support SSL/TLS or not.

For reference, on my ubuntu it shows this:

```
$  pkg-config hiredis --cflags
-D_FILE_OFFSET_BITS=64 -I/usr/include/hiredis
```

**Comment by joelbax** [comment]
@henningw  I finally got some time to work on this and used a similar approach of the one used in the Lua module. The problem is that hiredis is built with a Makefile flag and no macro is defined, so, I'm checking for SSL library existance and populate a WITH_SSL flag.

Unfortunately, after so much time, some indentation was fixed in this module and my fork had conflicts. Trying to fix them seems that screwed github's commit history. I'll cancel this, re-fork and re-apply my changes. 

I'm really sorry for all the nuisance.

**Pull request closed without merging**